### PR TITLE
ColorConversionXForm Module

### PR DESCRIPTION
**Pull request by mohammedzakikochargi**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #[Issue]

**Description**

This is color conversion module ,currently only for cpu based conversions. Both planar and raw conversions are a part of this module.

**Alternative(s) considered**

Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**

Type Choose one: Feature
**Screenshots (if applicable)**

**Checklist**

- [x] I have read the [Contribution Guidelines](https://github.com/Apra-Labs/ApraPipes/wiki/Contribution-Guidelines)
- [x] I have written Unit Tests
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach

### CODE REVIEW

File: base/CMakeLists.txt [hunk 1]
```diff
@@ -318,6 +321,8 @@ SET(IP_FILES_H
 	include/FaceDetectsInfo.h
 	include/FaceDetectorXform.h
 	include/TextOverlayXForm.h
+	include/ColorConversionXForm.h
+	include/AbsColorConversionFactory.h
```
Comment (kumaakh) [inline base/CMakeLists.txt:325]: is this factory required to be exposed via include or should this also be part of src ?

File: base/include/ColorConversionXForm.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+#include "FrameMetadata.h"
+#include "Module.h"
+#include "AbsColorConversionFactory.h"
+
+class ColorConversionProps : public ModuleProps
+{
+public:
+	enum ConversionType
+	{
+		RGB_2_MONO = 0,
```
Comment (kumaakh) [inline base/include/ColorConversionXForm.h:11]: The numeric '2' is conusing... it should br "TO"

File: base/src/ColorConversionStrategy.h [hunk 3]
```diff
@@ -0,0 +1,211 @@
+#include "ColorConversionXForm.h"
+
+class DetailAbstract
+{
+public:
+	DetailAbstract() {}
+	~DetailAbstract() {}
+	virtual void convert(frame_container& inputFrame, frame_sp& outFrame, framemetadata_sp outputMetadata) {};
```
Comment (kumaakh) [inline base/src/ColorConversionStrategy.h:8]: I dont understand why we need to remember iImg and oImg here ?
it is never a good idea to have unwanted state. It means one unwanted reference is always here.

File: base/src/ColorConversionXForm.cpp [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+bool ColorConversion::term()
+{
+	return Module::term();
+}
+
+bool ColorConversion::process(frame_container& frames)
+{
+	auto outFrame = makeFrame();
```
Comment (kumaakh) [inline base/src/ColorConversionXForm.cpp:157]: I like the concise ness of this function

File: base/src/Utils.cpp [hunk 5]
```diff
@@ -148,6 +149,12 @@ cv::Mat Utils::getMatHeader(RawImageMetadata* metadata)
 	return cv::Mat(metadata->getHeight(), metadata->getWidth(), metadata->getType(), static_cast<void*>(&data), metadata->getStep());
 }
 
+cv::Mat Utils::getMatHeader(RawImagePlanarMetadata* metadata)
+{
+	uint8_t data = 0;
+	return cv::Mat(metadata->getHeight(0) + metadata->getHeight(0) / 2, metadata->getWidth(0), CV_8UC1, static_cast<void*>(&data));
```
Comment (kumaakh) [inline base/src/Utils.cpp:155]: This is making an assumption of YUV420... not a good idea.

**Review by kumaakh — CHANGES_REQUESTED** [review]
Review comments are added. please imporve as suggested. 

Also, all the image files should be committed as LFS.

**Review by kumaakh — CHANGES_REQUESTED** [review]
Please address the review comments

File: base/CMakeLists.txt (hunk 1 above)
Comment (mohammedzakikochargi) [reply base/CMakeLists.txt]: Yes , it can be a part of src. Added it to src.

File: base/include/ColorConversionXForm.h (hunk 2 above)
Comment (mohammedzakikochargi) [reply base/include/ColorConversionXForm.h]: Done, changed it!!

File: base/src/ColorConversionStrategy.h (hunk 3 above)
Comment (mohammedzakikochargi) [reply base/src/ColorConversionStrategy.h]: Addressed the change.

File: base/CMakeLists.txt (hunk 1 above)
Comment (mohammedzakikochargi) [reply base/CMakeLists.txt]: Yes it was a mistake , i corrected it.

File: base/src/Utils.cpp (hunk 5 above)
Comment (mohammedzakikochargi) [reply base/src/Utils.cpp]: It was a mistake, i've resolved it.

**Review by kumaakh — CHANGES_REQUESTED** [review]

File: base/include/RawImageMetadata.h [hunk 6]
```diff
@@ -207,6 +216,15 @@ class RawImageMetadata : public FrameMetadata
 		case ImageMetadata::BGRA:
 			channels = 4;
 			break;
+		case ImageMetadata::BAYERBG10:
+			channels = 1;
```
[the remark this answers is not in the document]
Comment (mohammedzakikochargi) [reply base/include/RawImageMetadata.h]: Updated

File: base/include/ColorConversionXForm.h [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+	int mFrameType;
+	ColorConversionProps mProps;
+	boost::shared_ptr<DetailAbstract> mDetail;
+	std::string mOutputPinId;
+	uint16_t mWidth;
+	uint16_t mHeight;
+	framemetadata_sp mOutputMetadata;
+	framemetadata_sp mInputMetadata;
+	cv::Mat inpImg;
+	cv::Mat outImg;
```
[the remark this answers is not in the document]
Comment (mohammedzakikochargi) [reply base/include/ColorConversionXForm.h]: Resolved

File: base/src/AbsColorConversionFactory.cpp [hunk 8]
```diff
@@ -0,0 +1,111 @@
+#include "AbsColorConversionFactory.h"
+#include "ColorConversionStrategy.h"
+
+boost::shared_ptr<DetailAbstract> AbsColorConversionFactory::create(framemetadata_sp input, framemetadata_sp output, cv::Mat& inpImg, cv::Mat& outImg)
+{
+	boost::shared_ptr<DetailAbstract> mapper;
+	static std::map<std::pair<ImageMetadata::ImageType, ImageMetadata::ImageType>, boost::shared_ptr<DetailAbstract>> cache;
```
[the remark this answers is not in the document]
Comment (mohammedzakikochargi) [reply base/src/AbsColorConversionFactory.cpp]: Removed the cache.

File: base/src/ColorConversionStrategy.h [hunk 9]
```diff
@@ -0,0 +1,196 @@
+#include "ColorConversionXForm.h"
+
+class DetailAbstract
+{
+public:
+	DetailAbstract() {}
+	~DetailAbstract() {}
+	virtual void convert(frame_container& inputFrame, frame_sp& outFrame, framemetadata_sp outputMetadata, cv::Mat inpImg, cv::Mat outImg) {};
```
[the remark this answers is not in the document]
Comment (mohammedzakikochargi) [reply base/src/ColorConversionStrategy.h]: Removed unwanted param.

File: base/src/AbsColorConversionFactory.cpp [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+			mapper = boost::shared_ptr<DetailAbstract>(new CpuBayerGB82RGB());
+		}
+		else if (inputImageType == ImageMetadata::BAYERRG8 && outputImageType == ImageMetadata::RGB)
+		{
+			mapper = boost::shared_ptr<DetailAbstract>(new CpuBayerRG82RGB());
+		}
+		else if (inputImageType == ImageMetadata::BAYERGR8 && outputImageType == ImageMetadata::RGB)
+		{
+			mapper = boost::shared_ptr<DetailAbstract>(new CpuBayerGR82RGB());
+		}
```
[the remark this answers is not in the document]
Comment (mohammedzakikochargi) [reply base/src/AbsColorConversionFactory.cpp:63]: Resolved

File: base/src/AbsColorConversionFactory.cpp [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+		if (inputImageType == ImageMetadata::RGB && outputImageType == ImageMetadata::YUV420)
+		{
+			auto height = rawOutputMetadata->getHeight(0);
+			int outputRows = height * 1.5;
+
+			inpImg = Utils::getMatHeader(rawMetadata);
+			outImg = Utils::getMatHeader(rawOutputMetadata, outputRows);
+			mapper = boost::shared_ptr<DetailAbstract>(new CpuRGB2YUV420Planar());
+		}
```
[the remark this answers is not in the document]
Comment (mohammedzakikochargi) [reply base/src/AbsColorConversionFactory.cpp:80]: Added

File: base/src/AbsColorConversionFactory.cpp [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+		if (inputImageType == ImageMetadata::YUV420 && outputImageType == ImageMetadata::RGB)
+		{
+			auto height = rawMetadata->getHeight(0);
+			int inputRows = height * 1.5;
+
+			inpImg = Utils::getMatHeader(rawMetadata, inputRows);
+			outImg = Utils::getMatHeader(rawPlanarMetadata);
+			mapper = boost::shared_ptr<DetailAbstract>(new CpuYUV420Planar2RGB());
+		}
```
[the remark this answers is not in the document]
Comment (mohammedzakikochargi) [reply base/src/AbsColorConversionFactory.cpp:103]: Added

**Review by kumaakh — APPROVED** [review]

**Pull request merged**